modules/blocks/README.md

```diff
@@ -0,0 +1,162 @@
+
+- There's no built-in JavaScript MD5 hash function, so we pull in [a common implementation](https://github.com/blueimp/JavaScript-MD5).
+- When registering a new block, you must prefix its slug with a namespace for your plugin. This helps prevent conflicts when more than one plugin registers a block with the same slug.
+- You will use `createElement` to describe the structure of your block's markup. See the [Element documentation](../element/README.md) for more information.
+- Extracting `GravatarImage` to a separate function allows us to reuse it in both the editor-specific interface and the published content.
+- We only change the attributes of a block by calling the `setAttributes` helper. Never assign a value on the attributes object directly.
+
+There's a few small details that aren't obvious though:
+
+- How are a block's attributes stored when saving a post, so that they're restored next time I edit the same post?
+- Keen to avoid spam, how can I ensure I'm not including the email address in the published content?
+
+The single answer to both of these questions is that the combined state of every block in a post is saved as [post meta](https://codex.wordpress.org/Custom_Fields#Advanced_Techniques_for_Custom_Fields). This ensures that the attributes of a block are preserved and kept separate from the markup between editing sessions. 
+
+## API
+
+### `wp.blocks.registerBlock( slug: string, settings: Object )`
+
+Registers a new block provided a unique slug and an object defining its behavior. Once registered, the block is made available as an option to any editor interface where blocks are implemented.
+
+- `name: string` - A human-readable [localized](https://codex.wordpress.org/I18n_for_WordPress_Developers#Handling_JavaScript_files) label for the block. Shown in the block picker.
+- `edit( { attributes: Object, setAttributes: Function } ): WPElement` - Returns an element describing the markup of a block to be shown in the editor. A block can update its own state in response to events using the `setAttributes` function, passing an object of properties to be applied as a partial update.
+- `save( { attributes: Object } ): WPElement` - Returns an element describing the markup of a block to be saved in the published content. This function is called before save and when switching to an editor's HTML view.
+- `tagName: string` - An alternative to defining `edit` and `save` behaviors, if passed a [tag name](https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement#Parameters), delegates default behavior of an editable field on that node type. Typically used for simple, text-heavy elements (paragraphs, headings) to avoid excess meta storage of text.
+- `controls: string[]` - Slugs for controls to be made available to block. See also: [`wp.blocks.registerControl`](#wpblocksregistercontrol-slug-string-settings-object-)
+
+### `wp.blocks.registerControl( slug: string, settings: Object )`
```

Comment:
The tricky bit is that you might often want to make an image _linked_, but you rarely want to make it _bold_, whereas an image caption you'd often want to make both linked, and bold. 

We might end up with something close to what TinyMCE and Calypso has currently, a permanently visible toolbar. But let's not throw in the towel quite yet.

What if we did something like this? Image selected, most inline tools are grayed:

![image](https://cloud.githubusercontent.com/assets/1204802/24165368/7311c2ce-0e70-11e7-9cfa-43d39603f588.png)

When focus is on the caption:

![image captioning](https://cloud.githubusercontent.com/assets/1204802/24165431/a61f099c-0e70-11e7-8747-d04b9b3bf452.png)

In this mockup I unified the toolbar and separated with lines, to simplify the silhouette.